### DISCUSSION TITLE
[Rails 7] - Help to configure a newly migrated app to process activeadmin assets (js and scss) using jsbundling-rails with esbuild

### DISCUSSION THREAD

**Discussion by guillaumebesse** [activeadmin/activeadmin#7947] [Help]
Thanks to the work in progress in PR #7235, activeadmin is starting.

But I don't know how to deal with active_admin.js and my custom scss files. 
I am new to esbuild.

Any help is welcome.

**Comment by tagliala** [comment]
> But I don't know how to deal with active_admin.js and my custom scss files.
I am new to esbuild.

Hi

⚠️  Disclaimer: I'm not a JavaScript developer

I switched back to webpacker for the moment, because I can see that Active Admin [is using](https://github.com/activeadmin/activeadmin/blob/b732b2c239861b9a4200bb08349849f6ecb1b62d/app/javascript/active_admin/base.js#L2-L6) jQuery UI and this library does not play well with esbuild by default

I think this resource could be helpful: https://gorails.com/episodes/how-to-use-jquery-with-esbuild

It requires to download custom assets from jquery-ui and I'm not comfortable in doing that

Slightly refers to #7226

**Comment by guillaumebesse** [comment]
I did some progress, everything is not tested but It is really promising.

The goal is to compile in `app/assets/builds`
- two CSS files 
    - `application.css` for my web app and build with tailwindcss 
    -  `active_admin.css` build with sass compiler
- two js files
    - `application.js` for my web app 
    - `active_admin.js`  


Files tree
- app
    - assets
        - builds
        - config
            - `manifest.js` 
        - stylesheets
            - `active_admin.scss`
            - `application.tailwind.css`
    - javascript
        - controllers
            - `application.js`
            - `index.js`
        - `active_admin.js`
        - `application.js`    


1/ In package.json, I have

```
...
"scripts": {
    "build": "esbuild app/javascript/*.* --bundle --sourcemap --outdir=app/assets/builds",
    "build:css": "tailwindcss -i ./app/assets/stylesheets/application.tailwind.css -o ./app/assets/builds/application.css && sass ./app/assets/stylesheets/active_admin.scss ./app/assets/builds/active_admin.css --no-source-map --load-path=node_modules",
}
...
```
application.tailwind.css: I regroup all my CSS this file
```
@tailwind base;
@tailwind components;
@tailwind utilities;

// overriding default 
@layer base {
    h1 { @apply ...}
    h2 { @apply ...}
}


@layer components {
    .red_button { @apply whitespace-nowrap ... }
}
```

active_admin.scss
```
@import "@activeadmin/activeadmin/src/scss/mixins";
@import "@activeadmin/activeadmin/src/scss/base";

// my custom css
.yes {
  background-color: #c3d4ab;
}
```

active_admin.js
```
import "@activeadmin/activeadmin";
```

**Comment by j-manu** [comment]
I wanted to get this working with rails 7 and esbuild and took the nuclear option 😅 

1. Generate the final active_admin.js and active_admin.css in a new app.
2. Copy the generated files to the original app under app/assets/activeadmin 
3. Fork activeadmin and remove dependency on jquery-rails and remove app/assets, app/javascript and vendor/assets
4. Use the fork in the app

And it works!

It might have been possible to make it work with https://github.com/excid3/esbuild-rails but jquery-ui needs a custom build.

If activeadmin can publish a gem with assets precompiled, it will be much easier to use.

**Comment by Aesthetikx** [comment]
I have a similar issue, although it is entangled with my limited understanding of the new Rails 7 css and js options. I created a new Rails 7 app with `--tailwind`, and thusly have `tailwindcss-rails` in my `Gemfile`. Now, thanks to the works of the developers on the recent Rails 7 PR, I have added `activeadmin` using the `master` branch to my `Gemfile`.

After installing and visiting `/admin`, I am greeted with `The asset 'active_admin.css' was not found in the load path.` from `raise Propshaft::MissingAssetError.new(logical_path)`. Not sure where to go from here, or if I need to replace my tailwind install with one of the other css bundling options.

**Reply by JuzerShakir** [reply]
I also face the same issue using `propshaft` as an asset pipeline. However, when creating the app, I didn't specify the `jsbundling` & `cssbundling` options.

Did you find any solution to make the `activeadmin` work with `propshaft`?

**Comment by Kukunin** [comment]
I successfully manage to install `active_admin` with `vite_ruby` (ViteJS):

To handle the `undefined method assets` while running `rake activeadmin:install`, I temporary installed `gem 'propshaft'`. Later the gem is not needed. 

I've added the following snippet to mimic Webpacker methods the `config/initializers/active_admin.rb`:

```
module ActiveAdminViteJS
  def stylesheet_pack_tag(style, **options)
    style = 'active_admin.scss' if style == 'active_admin.css'
    vite_stylesheet_tag(style, **options)
  end

  def javascript_pack_tag(script, **options)
    vite_javascript_tag(script, **options)
  end
end

ActiveAdmin::Views::Pages::Base.include ActiveAdminViteJS
```

Ensure you have `config.use_webpacker = true` in the config.

`app/frontend/entrypoints/active_admin.js`:

```
import '../jquery'
import 'jquery-ui'
import 'jquery-ui/ui/widgets/mouse'
import '@activeadmin/activeadmin'
```

`app/frontend/entrypoints/active_admin.scss`:

```
@import "@activeadmin/activeadmin/src/scss/mixins";
@import "@activeadmin/activeadmin/src/scss/base";
```

`app/frontend/jquery.js`:

```
import $ from 'jquery'
window.jQuery = window.$ = $
```

No issues so far, but I guess jquery-ui components would require explicit importing

**Reply by jhirn** [reply]
This was super helpful! I also had to add this for favicons:  

```ruby
# in ActiveAdmin.config
config.favicon = 'images/favicon.png' 

# In ActiveAdminViteJS
def favicon_pack_tag(favicon)
  favicon_link_tag(vite_asset_path(favicon))
end
```

**Reply by danieldiekmeier** [reply]
Thank you, this really helped! However, I got an error:

```
NoMethodError:
  undefined method `stylesheet_pack_tag' for an instance of #<Class:0x00007f414a177770>
  /home/runner/work/flipvinyls/flipvinyls/vendor/bundle/ruby/3.3.0/gems/activeadmin-3.2.1/app/views/layouts/active_admin_logged_out.html.erb:10
```

I think this is running in another context because it is used for Devise. I was able to fix it by patching the `ActiveAdmin::ViewHelpers` instead. With this, all my pages work.

```diff
- ActiveAdmin::Views::Pages::Base.include ActiveAdminViteJS
+ ActiveAdmin::ViewHelpers.include ActiveAdminViteJS
```

**Comment by reinvanimschoot** [comment]
Hi guys, sorry to resurrect this discussion but while this solution works like a charm _inside_ Active Admin, I can't seem to get the sign in page working. I keep getting the following error:

```
ActionView::Template::Error (Shakapacker can't find active_admin.css in /Users/.../public/packs/manifest.json. Possible causes:

...

Your manifest contains:
{
}
```

I already removed most of the webpacker files and configs, but  I can't seem to find the place to make this particular page working, which means I can't log in to Active Admin at all.

Any help would be greatly appreciated!

**Comment by jedschneider** [comment]
Hi all, I've been back to this url several times in the last week, I wanted to report on my progress and what i think is a tenable solution for my needs. I started with the snippets from @guillaumebesse [up there ](https://github.com/activeadmin/activeadmin/discussions/7947#discussioncomment-5867899)☝️ . This was perfect to get me css / scss compiling and get me on the right track but as soon as i added inline forms that required javascript (has_many, nested attributes) is when I noticed the javascript was not getting built.

`active_admin.js`

```javascript
import "@activeadmin/activeadmin";
```

issue: web conole reports

> Uncaught SyntaxError: Cannot use import statement outside a module

solution (I think): declare the build to be a module in `package.json`

```json
  "module": "true",
  "dependencies": {
    "@activeadmin/activeadmin": "^3.0.0",
    ...
```

Issue: jquery is not defined:
solution: `import` statements hoist variables so you have to force jquery to be defined and added to the window in a separate file

`app/javascripts/add_jquery.js`

```javascript
import jquery from 'jquery';
window.jquery = jquery; // i mean why not?!
window.jQuery = jquery; // jquery ui is looking for this one
window.$ = jquery; // activeadmin wants this one
```

`application.js`

```javascript
import "./add_jquery" // forces hoisting to behave
import './jquery-ui' // jQuery is now defined
import "@activeadmin/activeadmin";
````

Final issue, jQuery UI when loaded from node_modules has a ton of load dependencies outside the initial widget.js, as you can see from the above import, I vendored jquery-ui instead of using it from the node modules. you can download the full version from the website. I finally came across [this video](https://www.youtube.com/watch?v=ql-Ng6OeI-M) which ended up being really useful to get that final step.

```shell
% tree app/javascript 
app/javascript
├── active_admin.js
├── add_jquery.js
├── application.js
├── controllers
│   ├── application.js
│   ├── hello_controller.js
│   └── index.js
└── jquery-ui.js
```

Like often, javascript has taken hours off of my life, but I'm super grateful to not be in the business of building forms, so thanks ActiveAdmin folks. Hopefully this helps streamline the process for others on Rails 7 and esbuild.

**Reply by jedschneider** [reply]
I discovered that the delete buttons/links in activeadmin were not working, this was an issue with `jquery-ujs` not being imported, so my `active_admin.js` now looks like this:

```javascript
import "./add_jquery"
import {} from 'jquery-ujs'
import './jquery-ui'
import "@activeadmin/activeadmin";
```

This allows the modal dialog to work and for the `data-confirm` attributes to be picked up on in that case.

**Reply by alexplatteeuw** [reply]
Hey @jedschneider, big thanks for dropping that comment! I was dealing with the same problem.

Initially, I tried importing turbo-rails into my `active_admin.js` file and then went on to replace all the `link_to` instances (that were using a patch/post/delete method and were not working anymore) with a button_to. I also fixed the modals by swapping out `data: { confirm: ` with `form: { data: { turbo_confirm:`.

After that, I took the time to tweak all my `button_to` to match the appearance of the `link_to`... 😑

Thanks to your comment I just realized that activeadmin is still relying on UJS through jquery-ujs.

In the end, I imported jquery-ujs like this in my `active_admin.js` file:

```js
import { } from "@activeadmin/activeadmin/node_modules/jquery-ujs"
```

**Reply by mario-amazing** [reply]
I've successfully configured ActiveAdmin version 3.2 in my Rails project with jsbundling-rails, and propshaft, all managed via node_modules. Here's the setup that worked for me:

```js
//app/javascript/active_admin.js

import './add_jquery'

import 'jquery-ui'
import "jquery-ujs"
import 'jquery-ui/ui/unique-id.js'
import 'jquery-ui/ui/widgets/mouse'
import "jquery-ui/ui/widgets/sortable"
import "jquery-ui/ui/widgets/tabs"
import "jquery-ui/ui/widgets/dialog"
import 'jquery-ui/ui/widgets/datepicker'
import 'jquery-ui/ui/widget'
import '@activeadmin/activeadmin'
```

**Reply by jhirn** [reply]
Dang I didn't follow the discussion from the original issue and wasted so much time with vite inject and other configuration trying to get jquery_ui to load properly. I ended up doing exactly this yesterday after losing a day on it. 

I believe the reason for this is the order on how Vite transforms CJS imports when `transformMixedEsModules: true` is set meaning `import jQuery from 'jquery'` will not actually set jQuery before transforming `import 'jQuery-ui` which depends on it being set. Moving the jQuery import to another file will load/set jQuery and also do your global assignments to `window.$` and `window.jQuery` if you have them.

**Reply by javierjulio** [reply]
This should work with jsbundling-rails and our NPM package as is. I think the issue may be that we aren't specifying the ESM file in a `"module": "..."` item in the `package.json` file. Can someone please confirm if removing the manual imports and just using `import '@activeadmin/activeadmin/src/base'` works for you? I'm pretty certain we just need to add that like other packages do, [for example, the ActionText gem package.json file](https://github.com/rails/rails/blob/v7.2.1/actiontext/package.json#L5-L6) as it should work without having to manually specify dependency imports.

**Reply by javierjulio** [reply]
It doesn't seem we are building a proper ESM version for v3. I'll try to see if we can [implement what ActionView did for rails-ujs](https://github.com/rails/rails/tree/v7.1.4/actionview) for bundling an additional ESM file with our rollup config.

**Reply by javierjulio** [reply]
I have a PR that adds a new `app/assets/javascripts/active_admin/base-esm.js` file that should help resolve this so you won't need to import each dependency and file manually. I would appreciate [any help in testing this new bundle file](https://github.com/activeadmin/activeadmin/pull/8497).

**Comment by aaoafk** [comment]
I'm not sure if this is necessarily still an issue for people but i was able to get this resolved with two steps. This seems to be acceptable considering the new propshaft asset model handling:

Roughly: transpile with dartsass-rails. That means running the set up for dartsass-rails and then i added a build step in `Procfile.dev` for `admin_css`

```
admin_css: bin/rails dartsass:watch
```

My initializer for `dartsass-rails` looks like the following:

```
# config/initializers/dartsass.rb
Rails.application.config.dartsass.builds = {
  "#{Rails.root}/vendor/assets/stylesheets/active_admin.scss" => "#{Rails.root}/vendor/assets/builds/active_admin.css"
}
```

You should do the above to make sure that your application css won't get clobbered by active_admin css. nothing else is really needed, propshaft will deliver the correct asset once it's transpiled for active admin css to use because active admin is using stylesheet_link_tag i believe

Then in the active_admin initializer you need to add this configuration:

```
 # clear any previous assets
  config.clear_javascripts!

  # We need jQuery and jQuery UI and jQuery-ujs which are dependencies
  config.register_javascript "https://cdn.jsdelivr.net/npm/jquery@3.6.0/dist/jquery.js"
  config.register_javascript "https://cdn.jsdelivr.net/npm/jquery-ujs@1.2.3/src/rails.min.js"
  config.register_javascript "https://cdn.jsdelivr.net/npm/jquery-ui@1.13.2/dist/jquery-ui.min.js"

  # The actual active admin javascript file
  config.register_javascript "https://cdn.jsdelivr.net/npm/@activeadmin/activeadmin@2.13.1/app/assets/javascripts/active_admin/base.min.js"
```

I had trouble actually figuring out the javascript. I tried esbuild to bundle the smaller files but that was leading to some other issues with imports and what not and seemed like a rabbit hole. Credit for the javascript goes to Henrik on medium: https://henrikbjorn.medium.com/til-using-activeadmin-with-importmap-rails-289cde6f76b8

He also has a section on importmaps but i don't fully understand importmaps yet so ...

anyways hope the above is helpful especially for people integrating activeadmin with rails 8. This seems to fully work, at least I tested one delete action and was able to successfully delete a resource on the admin backend.

**Comment by losnikitos** [comment]
**tldr**: To use Active Admin v3 with Propshaft, put https://gist.github.com/losnikitos/46b4fb8c63bfd14577299611ec90f61b into `app/assets/javascripts/active_admin.js` (quick and dirty solution)

---

I had a problem with missing js bundle after migrating from Sprockets to Propshaft (Rails 7,  Active Admin v3). My `active_admin.js` file was basically this one line
```
//= require active_admin/base
```
Which Sprockets would have picked up and assembled a bundle, but Propshaft wouldn't.

What I did was 'quick and dirty' solution: I assembled a bundle with sprockets (`rails assets:precompile`), grabbed `public/assets/active_admin-{digest}.js` and put the contents into `app/assets/javascripts/active_admin.js`. This would work if you don't plan to update active admin very often. Here's what the bundle looks for me (build for active admin v3.2.2) https://gist.github.com/losnikitos/46b4fb8c63bfd14577299611ec90f61b
